students/drovdahl/lesson08/circle.py

```diff
@@ -0,0 +1,74 @@
+#! /usr/local/bin/python3
+'''
+Class that does the following:
+    Compute the circle's area
+    Print the circle and get something nice
+    Add two circles together
+    Compare two circles to see which is bigger
+    Compare two circles to see if they are equal
+    Put circles in a list and sort them
+
+Use the following techniques
+    properties
+    classmethod
+    special methods
+'''
+import math
+
+
+class Circle(object):
+    def __init__(self, radius):
+        self._radius = radius
+        self.diameter = radius * 2
+
+    @property
+    def radius(self):
+        return self.diameter / 2
+
+    @property
+    def diameter(self):
+        return self._radius * 2
```

Comment:
Why not just `return self.diameter`?

Reply:
Comment from the peanut gallery:

The point of this exercise is that you only want to store either radius or diameter, but not both -- and calculate one from the other with a property when needed.

and in this case, you have defined a "diameter" property -- and then written over it with a "diameter" instance attribute. So the property will not be used at all.

Did this code work? In the sense of being able to change either radius or diameter and then have both still be correct?

Reply:
The code worked.  I'll refactor it and try it out the way you're suggesting as well.  thanks for the feedback!

Reply:
I see:

You have a radius property that is computing from diameter. so re-setting the diameter works.

but the self.diameter is over-riding the diameter property, so it isn't doing anything, and self._radius isn't used as a result.

I think you would find that you can't re-set the radius with this code -- there must not have been a test for that -- so hard to get comprehensive tests!

Reply:
You're ABSOLUTELY correct!  I just tried to reset the radius and failed...

Still not sure I completely understand this entirely but will spend some more cycles on this and try and get a handle on it.

Reply:
radius was a property with only a getter, and no setter, so you can't set it.

Reply:
Ah...  makes sense.  I might have missed where the assignment asked for a radius setter.

Reply:
it didn't -- it probably didn't specify, but it was expecting that radius would be a plain old attribute. But if you DO make it a property, you probaby want to provide a setter.

we should add that to the test suite.

Reply:
Oh, I see!  So, is it pretty much standard practice to configure both a setter and getter for any property you make?

Reply:
This is addressed in my next PR